### PR TITLE
Wip/hughsie/tpm event log replay

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

**Comment by superm1** [comment]
It's probably worth taking this the step further of comparing that the calculated hash _actually_ matches and making some noise if it doesn't.

**Comment by hughsie** [comment]
> comparing that the calculated hash actually matches

I guess we should do that in `tpm-eventlog` rather than `tpm` as the latter might be neutered in favour of the dell on, right? The `uefi` plugin isn't logically enough connected to the TPM, although maybe we can use the `registered()` event? Ideas welcome.

**Comment by superm1** [comment]
Looks like one of the CI problems is probably this fix:
https://gitlab.gnome.org/GNOME/glib/merge_requests/1307

We can probably turn off -Wtypedef-redefinition for the failing builds for it for now.

**Comment by superm1** [comment]
> I guess we should do that in tpm-eventlog rather than tpm as the latter might be neutered in favour of the dell on, right? The uefi plugin isn't logically enough connected to the TPM, although maybe we can use the registered() event? Ideas welcome.

I think it's most interesting in the uefi plugin.  Let it get calculated in tpm event log plugin and let uefi pick up a signal and make noise when there's a problem then and they don't add up.  That means that if you can't calculate event log (such as on Dell system that has secure boot turned off) you just don't send the signal to uefi plugin.

**Comment by hughsie** [comment]
> I think it's most interesting in the uefi plugin.

I think I implemented it the other way around, but I don't think it matters much.

**Comment by hughsie** [comment]
> With how scary this situation actually is

I don't know how scary it is. I guess the TPM could be doing the wrong thing in a non-scary way, but @mjg59 is the expert here.

> I think it's worth also adding this into the daemon log too perhaps.

I don't think anyone looks at the daemon log in reality. If it's really that serious we ought to add a new `FWUPD_DEVICE_FLAG_IS_BROKEN` and provide a translation in the various UIs.

**Comment by mjg59** [comment]
There are machines that only produce sha256 hashes, so I think you need to handle that case. I also haven't checked the backend code, so I don't know whether you handle TCG2 format event logs (which would be a prerequisite for the first point).

There are definitely firmwares out there that generate incomplete event logs and I think it's worth logging that, but right now I wouldn't recommend expressing that to the user until we have a reasonable idea of how common it is.

**Comment by hughsie** [comment]
> There are machines that only produce sha256 hashes, so I think you need to handle that case.

Right, will fix.

>  so I don't know whether you handle TCG2 format event logs

We do!

> There are definitely firmwares out there that generate incomplete event logs

Do you know why? e.g. out of space etc?

**Comment by mjg59** [comment]
> Do you know why? e.g. out of space etc?

The most common example I've seen is measuring the TXT ACM but failing to include it in the event log.

**Comment by hughsie** [comment]
@superm1 do you want to re-review given i've rewritten a lot of the string handling?

File: plugins/tpm-eventlog/fu-plugin-tpm-eventlog.c [hunk 1]
```diff
@@ -11,15 +11,30 @@
...
 
+struct FuPluginData {
+	GPtrArray		*pcr0s;
+};
+
 void
 fu_plugin_init (FuPlugin *plugin)
 {
+	fu_plugin_alloc_data (plugin, sizeof (FuPluginData));
+	fu_plugin_add_rule (plugin, FU_PLUGIN_RULE_RUN_BEFORE, "uefi");
```
Comment (superm1) [inline plugins/tpm-eventlog/fu-plugin-tpm-eventlog.c]: Is this something that we should concern turning into a circular dependency?
UEFI has this:
```
        fu_plugin_add_rule (plugin, FU_PLUGIN_RULE_METADATA_SOURCE, "tpm_eventlog");
```

Comment (hughsie) [reply plugins/tpm-eventlog/fu-plugin-tpm-eventlog.c]: > Is this something that we should concern turning into a circular dependency

Good question. The `METADATA_SOURCE` rule doesn't affect plugin startup order and so this should be fine.

**Comment by superm1** [comment]
Just as an FYI, I did confirm on an XPS 7390 with secure boot enabled that this is working properly:
``
17:39:30:0409 FuPluginTpmEventlog  TPM reconstructed event log matched PCR0 reading
```

**Pull request merged**